Brokerages/Zerodha/ZerodhaBrokerage.cs

```diff
@@ -0,0 +1,1049 @@
+
+        private void OnError(object sender, WebSocketError e)
+        {
+            Log.Error($"ZerodhaSubscriptionManager.OnError(): Message: {e.Message} Exception: {e.Exception}");
+        }
+
+        private void OnMessage(object sender, MessageData e)
+        {
+            LastHeartbeatUtcTime = DateTime.UtcNow;
+
+            // Verify if we're allowed to handle the streaming packet yet; while we're placing an order we delay the
+            // stream processing a touch.
+            try
+            {
+                if (_streamLocked)
+                {
+                    _messageBuffer.Enqueue(e);
+                    return;
+                }
+            }
+            catch (Exception err)
+            {
+                Log.Error(err);
+            }
+
+            OnMessageImpl(e);
+        }
+
+        /// <summary>
+        /// Implementation of the OnMessage event
+        /// </summary>
+        /// <param name="e"></param>
+        private void OnMessageImpl(MessageData e)
+        {
+            try
+            {
+                if (e.MessageType == WebSocketMessageType.Binary)
+                {
+                    if (e.Count > 1)
+                    {
+                        int offset = 0;
+                        ushort count = ReadShort(e.Data, ref offset); //number of packets
+
+                        for (ushort i = 0; i < count; i++)
+                        {
+                            ushort length = ReadShort(e.Data, ref offset); // length of the packet
+                            Messages.Tick tick = new Messages.Tick();
+                            if (length == 8) // ltp
+                                tick = ReadLTP(e.Data, ref offset);
+                            else if (length == 28) // index quote
+                                tick = ReadIndexQuote(e.Data, ref offset);
+                            else if (length == 32) // index quote
+                                tick = ReadIndexFull(e.Data, ref offset);
+                            else if (length == 44) // quote
+                                tick = ReadQuote(e.Data, ref offset);
+                            else if (length == 184) // full with marketdepth and timestamp
+                                tick = ReadFull(e.Data, ref offset);
+                            // If the number of bytes got from stream is less that that is required
+                            // data is invalid. This will skip that wrong tick
+                            if (tick.InstrumentToken != 0 && offset <= e.Count && tick.Mode == Constants.MODE_FULL)
+                            {
+                                var symbol = _symbolMapper.ConvertZerodhaSymbolToLeanSymbol(tick.InstrumentToken);
+                                EmitQuoteTick(symbol, tick.Bids, tick.BuyQuantity, tick.Offers, tick.SellQuantity);
+                                EmitTradeTick(symbol, tick.LastTradeTime.GetValueOrDefault(), tick.LastPrice, tick.LastQuantity);
```

Comment:
Calling both `EmitQuoteTick` and `EmitTradeTick` would re-emit the last trade on every bid/ask changes?

- These two methods should be called in separate code paths
- `tick.LastTradeTime.GetValueOrDefault()` -- the timestamp of a Trade tick cannot be null
- The tick timestamp (for both trades and quotes) should be first converted from UTC to ExchangeTimeZone and them passed into these methods
  (e.g. OandaBrokerage: https://github.com/QuantConnect/Lean/blob/master/Brokerages/Oanda/OandaRestApiV20.cs#L418-L427)

Reply:
At the moment, Zerodha provides everything is a single WebSocket packet as mentioned here : https://kite.trade/docs/connect/v3/websocket/#quote-packet-structure for both QuoteTicks and TradeTicks. And for the timestamp part Zerodha / Exchange timestamp is already unix timestamp with timezone support when converted (We verified this already by running lean from Ireland with the help of @aman-ray.